### ISSUE TITLE
eselect alternative is totally broken on prefix

### ISSUE THREAD

**Issue by kiwifb** [gentoo-science/sci#453]
In the past I have fixed eselect to work on prefix. The fix doesn't apply anymore and trying to port it as is doesn't work either because of changes in alternatives-2.eclass.

Typical problem with the current set up on prefix. links and files normally installed under `/etc/env.d/alternatives` are not installed in the right place. They are installed in so called double prefix. So you find them under `$EPREFIX/$EPREFIX/etc/env.d/alternatives` that's one `$EPREFIX` too many.

Root cause are found in `usr/share/eselect/libs/alternatives-common.bash`, namely
```
inherit config output path-manipulation

: "${ALTERNATIVESDIR_ROOTLESS:=/etc/env.d/alternatives}"
ALTERNATIVESDIR="${EROOT%/}${ALTERNATIVESDIR_ROOTLESS}"
```
In a prefix because `ALTERNATIVESDIR_ROOTLESS` is populated with ` @sysconfdir@/env.d/alternatives` the end result will include the prefix as well. So in a prefix you will have
```
inherit config output path-manipulation

: "${ALTERNATIVESDIR_ROOTLESS:=${EPREFIX}/etc/env.d/alternatives}"
ALTERNATIVESDIR="${EROOT%/}${ALTERNATIVESDIR_ROOTLESS}"
```
where `$EPREFIX` is expanded to its value. `$EROOT` is equivalent to `$ROOT/$EPREFIX` so `ALTERNATIVESDIR` ends up with a double prefix. Bad.

The old solution that I created before `ALTERNATIVESDIR*` where put in `alternatives-common.bash` and was still in `alternatives.bash` was to remove the extra prefix from `ALTERNATIVESDIR_ROOTLESS` with `sed` during `src_install`.

This is unfortunately rendered moot by the following in `alternatives-2.eclass
```
alternatives_for() {
	debug-print-function ${FUNCNAME} "${@}"

	dodir /etc/env.d/alternatives

	ALTERNATIVESDIR_ROOTLESS="${ED}/etc/env.d/alternatives" \
		eselect alternatives add ${@} || die

	ALTERNATIVES_CREATED+=( ${1} )
}
```
Again we have `$ED` which is `$D/$EPREFIX` again `$D` is already including the prefix so affixing `$EROOT` will double it. And you end creating stuff in strange place during `src_install`.

The most easy fix is to define
```
ALTERNATIVESDIR="${ROOT%/}${ALTERNATIVESDIR_ROOTLESS}"
```
that will install things in the right place but further work is needed as the link are not pointing to the right place. They go to what would be the right location on a non-prefix system.

**Comment by kiwifb** [comment]
I will make a push request soon. The short of it: `EROOT` is banned in the alternatives scripts. `EPREFIX` must be added when setting the alternatives in the ebuild. This will be a massive pull....

**Comment by kiwifb** [comment]
I will amend that, I may be able to avoid changing any ebuilds by adding a single `EROOT` in a place there was none before. Experimenting.

**Comment by kiwifb** [comment]
`eselect ... update` needs love too.

**Comment by kiwifb** [comment]
The `update` is out of scope for this, this is currently in broken state prefix or not.

**Comment by heroxbd** [comment]
I can reproduce this bug.

**Comment by kiwifb** [comment]
There is stuff in #465 but jlec made changes that may require rebasing, but I have been working with that PR on prefix for a while.

**Comment by heroxbd** [comment]
ALTERNATIVESDIR_ROOTLESS="${ED}/etc/env.d/alternatives" is not a good idea.

I don't think eselect has access to ${D} or ${ED}. 

But alternatives-common.bash has to know ${D} in order to get relative paths correctly (line 309).  Therefore this fails on native Gentoo as well.

**Comment by kiwifb** [comment]
Note that the bit of code I quote using `${ED}` is in the eclass not `eselect`.

**Comment by junghans** [comment]
Yup, it is broken on fresh native installs, too!

**Comment by kiwifb** [comment]
I guess it's improved as I don't have a double prefix anymore but there is still a bit missing
```
>>> Installing (1 of 1) sci-libs/blas-reference-20151113::science
 * checking 7 files for package collisions
>>> Merging sci-libs/blas-reference-20151113 to /
--- /home/frb15/gentoo/etc/
--- /home/frb15/gentoo/etc/env.d/
--- /home/frb15/gentoo/etc/env.d/alternatives/
--- /home/frb15/gentoo/etc/env.d/alternatives/blas/
>>> /home/frb15/gentoo/etc/env.d/alternatives/blas/reference/
>>> /home/frb15/gentoo/etc/env.d/alternatives/blas/reference/usr/
>>> /home/frb15/gentoo/etc/env.d/alternatives/blas/reference/usr/lib/
>>> /home/frb15/gentoo/etc/env.d/alternatives/blas/reference/usr/lib/pkgconfig/
>>> /home/frb15/gentoo/etc/env.d/alternatives/blas/reference/_importance
--- /home/frb15/gentoo/usr/
--- /home/frb15/gentoo/usr/share/
--- /home/frb15/gentoo/usr/share/doc/
>>> /home/frb15/gentoo/usr/share/doc/blas-reference-20151113/
>>> /home/frb15/gentoo/usr/share/doc/blas-reference-20151113/README.bz2
--- /home/frb15/gentoo/usr/lib/
>>> /home/frb15/gentoo/usr/lib/librefblas.so.3.6.0
--- /home/frb15/gentoo/usr/lib/pkgconfig/
>>> /home/frb15/gentoo/usr/lib/pkgconfig/refblas.pc
>>> /home/frb15/gentoo/usr/lib/librefblas.so.3 -> librefblas.so.3.6.0
>>> /home/frb15/gentoo/usr/lib/librefblas.so -> librefblas.so.3
 * QA Notice: Symbolic link /home/frb15/gentoo/etc/env.d/alternatives/blas/reference/usr/lib/pkgconfig/blas.pc points to /home/frb15/usr/lib/pkgconfig/refblas.pc which does not exist.
>>> /home/frb15/gentoo/etc/env.d/alternatives/blas/reference/usr/lib/pkgconfig/blas.pc -> ../../../../../../../../../../../../../../../../../../usr/lib/pkgconfig/refblas.pc
>>> sci-libs/blas-reference-20151113 merged.
>>> Auto-cleaning packages...
```
Ideally it should have been `/home/frb15/gentoo/usr/lib/pkgconfig/refblas.pc` the relative link goes up one level too many.

**Comment by heroxbd** [comment]
Thanks François.  Please test https://github.com/gentoo-science/sci/pull/541.  It works for me.

**Comment by kiwifb** [comment]
Actually trying that right now.

**Comment by kiwifb** [comment]
Works for me. Did it twice to make sure. There are unusual verbose bits
```
>>> Safely unmerging already-installed instance...
+ for alt in '${ALTERNATIVES_CREATED[@]}'
+ eselect alternatives update blas
+ ret=0
+ case ${ret} in
+ :
+ set +x
```

**Comment by heroxbd** [comment]
The verbose output was accidentally introduced in a squashed commit (bc36b5406f2bb289)  and removed later (ecdd496c665cd).  Don't worry about that.  It will disappear once this PR is merged.

Thanks a lot for your input.
